funnel/views/post.py

```diff
@@ -0,0 +1,208 @@
+            else []
+        ),
+    }
+
+
+@Project.views('updates')
+@route('/<profile>/<project>/updates')
+class ProjectPostView(ProjectViewMixin, UrlForView, ModelView):
+    __decorators__ = [legacy_redirect]
+
+    @route('', methods=['GET'])
+    @render_with('project_posts.html.jinja2', json=True)
+    @requires_roles({'reader'})
+    def posts(self):
+        project_save_form = SavedProjectForm()
+        return {
+            'posts': self.obj.views.json_posts(),
+            'project': self.obj.current_access(),
+            'new_post': self.obj.url_for('new_post'),
+            'project_save_form': project_save_form,
+            'csrf_form': forms.Form(),
+        }
+
+    @route('new', methods=['GET', 'POST'])
+    @requires_login
+    @requires_roles({'editor'})
+    def new_post(self):
+        post_form = PostForm()
+        post_form.form_nonce.data = post_form.form_nonce.default()
+
+        if post_form.validate_on_submit():
+            post = Post(user=current_auth.user, project=self.obj)
+            post_form.populate_obj(post)
+            post.name = make_name(post.title)
+            db.session.add(post)
+            db.session.commit()
+            if post_form.restricted.data:
+                post.make_restricted()
+            db.session.commit()
+            return redirect(post.url_for(), code=303)
+
+        return render_form(
+            form=post_form,
+            title=_("Post an update"),
+            submit=_("Save & preview"),
+            cancel_url=self.obj.url_for(),
+        )
+
+
+ProjectPostView.init_app(app)
+
+
+@Post.features('publish')
+def post_publishable(obj):
+    return obj.state.DRAFT and 'editor' in obj.roles_for(current_auth.user)
+
+
+@Post.views('main')
+@route('/<profile>/<project>/updates/<url_name_uuid_b58>')
+class ProjectPostDetailsView(UrlForView, ModelView):
+    __decorators__ = [legacy_redirect]
+    model = Post
+    route_model_map = {
+        'profile': 'project.profile.name',
+        'project': 'project.name',
+        'url_name_uuid_b58': 'url_name_uuid_b58',
+    }
+
+    def loader(self, profile, project, url_name_uuid_b58):
+        project = (
+            Project.query.join(Profile)
+            .filter(Profile.name == profile, Project.name == project)
+            .one_or_404()
+        )
+
+        post = (
+            self.model.query.join(Project)
+            .filter(
+                Project.id == project.id, Post.url_name_uuid_b58 == url_name_uuid_b58,
+            )
+            .one_or_404()
+        )
+
+        if post.visibility_state.RESTRICTED and not (
+            project.current_roles.participant or project.current_roles.crew
+        ):
+            abort(403)
+
+        g.profile = post.project.profile
+        return post
+
+    @route('', methods=['GET'])
+    @render_with('project_post_details.html.jinja2')
+    @requires_roles({'reader'})
+    def view(self):
+        project_save_form = SavedProjectForm()
```

Comment:
Why does the Post view need to be aware of what forms the Project template requires? This should be coming out of a base class. Have a `ProjectFeatureMixin` or `ProjectHeaderMixin` class that provides a method that all these views can call. Pass these to the template namespaced: `'project_forms': self.project_header_forms()`. Let the template render using `project_forms.save` instead of `project_save_form`.